src/clients/twitch.ts

```diff
@@ -12,54 +15,76 @@ function registerTwitch(): void
+        }
+        catch(error: unknown)
+        {
+            const res: AuthResponse = await authenticateTwitch({
+                scope: configFile.T2D_SCOPE,
+                redirect_uri: configFile.T2D_REDIRECT_URI,
+                client_id: configFile.T2D_CLIENT_ID,
+                client_secret: configFile.T2D_SECRET,
+                use_https: configFile.T2D_HTTPS.enabled
             });
 
-        bridge.twitchClient = tmijs.Client({
-            options: { debug: true },
-            identity: {
-                username: configFile.T2D_USER,
-                password: 'oauth:' + access_token
-            },
-            channels: configFile.T2D_CHANNELS
-        });
+            await fs.writeFile('./tokens.json', JSON.stringify(res));
 
-        bridge.twitchClient.connect().then(() =>
-        {
-            console.log('Twitch bot is live! Sending a buffer message...', configFile.T2D_USER);
-            //Send a buffer message that allows us to track messages being sent as the twitch bot
-            //TODO: if we need to scale this to *much* more than just one twitch channel, this won't be usable, there will need to be another approach to record the ID's of the bot user
-            bridge.twitchClient!.say(configFile.T2D_CHANNELS[0], 'twitch bot buffer message!');
-        });
+            loginToTwitch();
+            return;
+        }
 
-        bridge.twitchClient.on('message', (channel: string, userState: tmijs.Userstate, msg: string, self: boolean) =>
-        {
-            //Send a message to twitch
-            //Something fascinating is, if a message is sent by the bot, self will be true. If I use the same username though when I chat, it's false.
-            if(!self)
+        // If bad data is given to our auth provider you'll get a log along the lines of "no valid token avaiable; trying to refresh.." etc.
+        const authProvider: RefreshingAuthProvider = new RefreshingAuthProvider(
             {
-                // Good idea to check if it exists or not, dunno.
-                if(!bridge.targetDiscordChannel)
-                {
-                    throw new Error('Cannot find Discord channel.');
-                }
+                'clientId': configFile.T2D_CLIENT_ID,
+                'clientSecret': configFile.T2D_SECRET,
+                onRefresh: async newTokenData => await fs.writeFile('./tokens.json', JSON.stringify(newTokenData, null, 4), 'utf-8')
+            },
+            tokenData! as any || await authenticateTwitch({
+                scope: configFile.T2D_SCOPE,
+                redirect_uri: configFile.T2D_REDIRECT_URI,
+                client_id: configFile.T2D_CLIENT_ID,
+                client_secret: configFile.T2D_SECRET,
+                use_https: configFile.T2D_HTTPS.enabled
+            })
+        );
+
+        bridge.twitch.authChatClient = new ChatClient({ authProvider, channels: [configFile.T2D_CHANNELS[0]] });
```

Comment:
So it looks like there are two clients, one that has authentication or logged in as a user, and one that doesn't. Based on what I'm reading the anon client is taking in messages, while the auth is sending them

My question is how come we separate them in the first place? Doing this increases the workload at the very least memory wise.

An explanation above line 61 in the code itself would probably be useful here too, since this might be a little odd at first to read :mag:

Reply:
If we listen for messages with an authed ChatClient we will not receive the event when we call ChatClient#say, so we have to create another ChatClient which is not authed/doesn't have the say method called on it to recieve the event.